### PR TITLE
add tracing to evaluate

### PR DESCRIPTION
**Pull request by jessechancy**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Add tracing to mlflow.evaluate() with support for regular python functions, Pyfunc, and langchain

When a user passes a chain or function to mlflow.evaluate() that has been instrumented with MLflow tracing, MLflow should log traces during evaluation and link them to the evaluation run using the same tagging approach from https://databricks.atlassian.net/browse/ML-40783 . 

We should make sure to support LangChain: even if the user hasn’t enabled LangChain autologging, mlflow.evaluate() should enable autologging for the duration of the evaluation so that traces are captured.

We should make sure to only enable the tracing part of LangChain autologging, not the part that logs models

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/models/evaluation/base.py
Comment (liangz1) [inline mlflow/models/evaluation/base.py:2046]:
```suggestion
    if model is not None and hasattr(model, "predict") and callable(model.predict):
```

**Review by liangz1 — COMMENTED** [review]
Looks good so far! Shall we add a test?

**Comment by BenWilson2** [comment]
The tracing branch is now closed. Can you rebase with a merge target to master?

File: tests/evaluate/test_evaluation.py
Comment (liangz1) [inline tests/evaluate/test_evaluation.py:395]: What does the trace look like with a regressor model? In which scenario would the user want to see a trace for a regressor evaluation?

File: tests/evaluate/test_evaluation.py
Comment (liangz1) [inline tests/evaluate/test_evaluation.py:382]: Shall we also assert that the traces are associated with the eval run?

**Review by liangz1 — COMMENTED** [review]
Left some questions about clarifying the supported functionality. Thanks!

File: tests/evaluate/test_evaluation.py
Comment (jessechancy) [reply tests/evaluate/test_evaluation.py:366]: 1. All models types should be supported. Anything other than langchain would just trace the input/output. In langchain, it would create a span for each individual section in langchain.
2. The traces would appear in the traces section as if you were to log a regular trace 
3. I think there isn't a difference between a pyfunc model or function here, because `mlflow.evaluate` does the neccessary steps to turn it all into a `model` with a `predict` method. Here we would be tracking the input/output of this predict method.

File: mlflow/utils/autologging_utils/__init__.py
Comment (BenWilson2) [inline mlflow/utils/autologging_utils/__init__.py:517]:
```suggestion
        yield
```

File: tests/evaluate/test_evaluation.py
Comment (BenWilson2) [inline tests/evaluate/test_evaluation.py:465]: We may want to add one final test to this suite. 
1. Call evaluate on a model
2. Verify traces are recorded for a simple langchain chain
3. Call mlflow autolog (specifying no overrides to the default) 
4. Invoke the model 
5. Ensure that the default settings for autologging apply and that the expected tracked run and artifacts / metrics / params are logged to the run. 

Mostly to ensure that modifications to autologging behavior within a single sessions are locally scoped with the configuration overrides being used in the context handler (it works as expected now, but in the future, if this logic needs to change, having the test will be a guard against a regression in the behavior introduced in this PR)

**Review by BenWilson2 — APPROVED** [review]
LGTM once the safeguard test is added

File: tests/evaluate/test_evaluation.py
Comment (liangz1) [reply tests/evaluate/test_evaluation.py:405]: Hi @jessechancy , is this addressed somewhere? Could you share a pointer if possible?

Besides, I'm wondering if we should provide a config to turn off the tracing for evaluate() runs, in case non-genai users get confused with the auto-logged traces. cc @BenWilson2

**Review by liangz1 — APPROVED** [review]
LGTM overall, just a question about the non-genai eval run case.
@jessechancy Do we have another JIRA ticket or PR for updating the documentation? If so, could you link it? Thanks!

File: mlflow/models/evaluation/base.py
Comment (liangz1) [inline mlflow/models/evaluation/base.py:2074]: Shall we check whether langchain is imported? I see this code block is not langchain-specific, so if the customer only uses non-langchain packages, shall we skip this?

File: .github/workflows/master.yml
Comment (liangz1) [inline .github/workflows/master.yml:243]: Do we want to make `langchain` a common deps? It seems to me that langchain should not be required for non-genai users. (related to my comment in `mlflow/models/evaluation/base.py`)

Can we test if langchain is not installed, a user should still be able to call `mlflow.evaluate()`?

**Review by liangz1 — CHANGES_REQUESTED** [review]
I noticed some langchain deps changes that might interrupt non-langchain users. Can you help make sure the experience is still smooth in that case? Thanks!

**Comment by BenWilson2** [comment]
Regarding https://github.com/mlflow/mlflow/pull/11911#discussion_r1599003847 - this is valid. I feel like the non-genai use cases (classification, regression) should be opt-in for this behavior as it doesn't make a great deal of sense and could potentially generate a very large dataset that doesn't provide any value, even for debugging.

File: mlflow/models/evaluation/base.py
Comment (BenWilson2) [reply mlflow/models/evaluation/base.py:2074]: +1 we might want to only apply this enablement if langchain core module is available through a validation with importlib

File: .github/workflows/master.yml
Comment (BenWilson2) [reply .github/workflows/master.yml:243]: It should not be a common dependency - this line change is only for test infra. 
That being said, a +1 to the validation check that the patch function with evaluate will work if langchain is not installed. Temporarily removing the install from the REPL context would be good to validate.

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:2077]: This is quite dangerous since it mutates the `model` object supplied by the user. It also doesn't support the case where an env manager is used to restore the model.

File: mlflow/utils/autologging_utils/__init__.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/__init__.py:519]: @jessechancy Can we specify all intended values of all keyword arguments here to ensure that we don't accidentally log models and other things?

E.g. `log_models=False`, etc.

Comment (dbczumar) [inline mlflow/utils/autologging_utils/__init__.py:519]: @serena-ruan Shouldn't we have a `log_traces` flag and remove the `log_inputs_outputs` flag since we no longer need to log inputs & outputs as a DataFrame?

@jessechancy While we're here, perhaps we can make that change?

File: mlflow/utils/autologging_utils/__init__.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/__init__.py:54]:
```suggestion
# Exempts autologging flavors from the `_AUTOLOGGING_GLOBALLY_DISABLED` flag
```

File: mlflow/utils/autologging_utils/__init__.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/__init__.py:55]:
```suggestion
_AUTOLOGGING_GLOBALLY_DISABLED_EXEMPTIONS = []
```

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:65]: This seems to be unused. Can we move it to https://github.com/mlflow/mlflow/pull/11911/files#r1602160113?

File: tests/evaluate/test_evaluation.py
Comment (dbczumar) [reply tests/evaluate/test_evaluation.py:529]: Great idea. Can we also test:

1. Call autolog with some specific settings
2. Call evaluate on a model
3. Verify traces are recorded for a simple langchain chain
4. Invoke the model
5. Ensure that the settings for autologging in (1) apply and that the expected tracked run and artifacts / metrics / params are logged to the run.

Can we also test these cases when evaluate fails with an exception?

File: mlflow/utils/autologging_utils/__init__.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/__init__.py:523]:
```suggestion
        try:
            mlflow.langchain.autolog(log_inputs_outputs=False)
            yield
        finally:
            mlflow.langchain.autolog(**prev_langchain_params)
```

File: tests/evaluate/test_evaluation.py
Comment (liangz1) [inline tests/evaluate/test_evaluation.py:508]: Looks good!

**Review by liangz1 — COMMENTED** [review]
LGTM with the non-langchain case. Thanks @jessechancy !

**Review by liangz1 — APPROVED** [review]

File: mlflow/utils/autologging_utils/__init__.py
Comment (jessechancy) [reply mlflow/utils/autologging_utils/__init__.py]: Can we do this in a followup PR? There are some discussions related to how this can be backwards compatible, and what to gate behind which flag / default values that might block this PR

File: mlflow/models/evaluation/base.py
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1268]: @jessechancy This will break all existing third-party evaluator plugins, including ones developed by Databricks. Can we construct `model_predict_func` inside the default evaluator instead?

**Review by dbczumar — CHANGES_REQUESTED** [review]
https://github.com/mlflow/mlflow/pull/11911/files#r1609573086

File: mlflow/utils/autologging_utils/__init__.py
Comment (dbczumar) [reply mlflow/utils/autologging_utils/__init__.py:530]: @jessechancy can we address this nit?

File: mlflow/utils/autologging_utils/__init__.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/__init__.py:529]: Is it sufficient to check `sys.modules` instead? Are there examples that recommend using `find_spec()` instead?

File: mlflow/models/evaluation/base.py
Comment (ian-ack-db) [reply mlflow/models/evaluation/base.py:1268]: +1 that it seems naturally scoped to `_extract_predict_fn` except in the case that a user would pass this explicitly into a DefaultEvaluator. Is there a clear use case around that we are trying to address?

File: mlflow/models/evaluation/base.py
Comment (ian-ack-db) [inline mlflow/models/evaluation/base.py:2069]: This comment `Disable all autologging except for traces` should more naturally go to the function `restrict_langchain_autologging_to_traces_only` itself. Right now I don't think this comment adds much over the name of the function. However, some discussion in that doc string around care taken to ensure non-langchain use cases are not broken would be nice

File: mlflow/utils/autologging_utils/__init__.py
Comment (ian-ack-db) [reply mlflow/utils/autologging_utils/__init__.py:66]: Should this have a name closer to `restrict_langchain_autologging_to_traces_only` which conveys better the intention to restrict logging to traces only. (As someone unfamiliar with autlogging I don't see traces really present in this config for how that's restricted).

Comment (jessechancy) [reply mlflow/utils/autologging_utils/__init__.py:66]: Where did you find the `extra_model_classes` parameter? I don't see it on `mlflow.langchain.autolog`

Comment (ian-ack-db) [reply mlflow/utils/autologging_utils/__init__.py:66]: https://github.com/mlflow/mlflow/pull/12022

Comment (ian-ack-db) [reply mlflow/utils/autologging_utils/__init__.py:66]: But this gets to a general point of how we will keep these in sync since this is a recent change it seems.

File: mlflow/models/evaluation/base.py
Comment (jessechancy) [reply mlflow/models/evaluation/base.py:1268]: I think the `_extract_predict_fn` doesn't cover all cases here, since I see direct usage of `model.predict`

File: mlflow/utils/autologging_utils/__init__.py
Comment (jessechancy) [reply mlflow/utils/autologging_utils/__init__.py:529]: There are a few examples throughtout the codebase that use `importlib.util.find_spec`, but I can change it to `sys.modules`

File: mlflow/utils/autologging_utils/__init__.py
Comment (jessechancy) [reply mlflow/utils/autologging_utils/__init__.py]: great point! I've added a test case to make sure it is in sync

**Review by ian-ack-db — COMMENTED** [review]
LGTM, though not familiar enough with implications for official approval

File: mlflow/models/evaluation/default_evaluator.py
Comment (dbczumar) [inline mlflow/models/evaluation/default_evaluator.py:145]: Can we add a docstring that indicates the precedence order and restructure this method so that the logic is if / elif / else? It's currently difficult to reason about what happens when `model_predict_fn` is specified and `raw_model` is also specified, etc.

Can we also add type hints to the function to indicate which parameters are optional?

File: tests/evaluate/test_evaluation.py
Comment (dbczumar) [reply tests/evaluate/test_evaluation.py:377]: > Anything other than langchain would just trace the input/output

It's up to the user to decide what level of granularity to use when tracing; if the user instruments each subroutine in their model with @mlflow.trace, then they won't just get overall input / output. We're not limiting / restricting this granularity in this PR, right @jessechancy ?

> The traces would appear in the traces section as if you were to log a regular trace

The key point is that we set the run ID as a tag on the trace, which appears in the UI.

> I think there isn't a difference between a pyfunc model or function here, because mlflow.evaluate does the neccessary steps to turn it all into a model with a predict method. Here we would be tracking the input/output of this predict method.

Same as above - the user might instrument at a level that's more granular than just input / output. Can we confirm this works?

**Review by dbczumar — APPROVED** [review]
LGTM once small comments are addressed. Thanks @jessechancy !

**Comment by heyitsaamir** [comment]
@jessechancy, I'm trying to log traces for my open ai function inside evaluate. Was there a reason only langchain was exempted here? Trying to gather data before deciding if I need to log an issue or add a fix

**Comment by heyitsaamir** [comment]
> @heyitsaamir There is an [open issue](https://github.com/mlflow/mlflow/issues/12952) and we are planning to prioritize it in a near future release. Feel free to add a thumb up in that issue!

Ah looks like I created a dupe of that bug possibly - https://github.com/mlflow/mlflow/issues/13765

**Pull request merged**
